### PR TITLE
Replace to snapshot applier

### PR DESCRIPTION
**Pull request by esuwu**

### CODE REVIEW

File: pkg/state/transaction_checker_test.go [hunk 1]
```diff
@@ -22,19 +22,43 @@ var (
...
+			assets:            stor.entities.assets,
+			scriptsStorage:    stor.entities.scriptsStorage,
+			scriptsComplexity: stor.entities.scriptsComplexity,
+			sponsoredAssets:   stor.entities.sponsoredAssets,
+			ordersVolumes:     stor.entities.ordersVolumes,
+			accountsDataStor:  stor.entities.accountsDataStor,
+			leases:            stor.entities.leases,
+		},
+	)
+	snapshotGen := snapshotGenerator{stor: stor.entities, scheme: settings.MainNetSettings.AddressSchemeCharacter}
```
Comment (nickeskov) [inline pkg/state/transaction_checker_test.go:55]: Use constructor here.

**Review by nickeskov — CHANGES_REQUESTED** [review]

File: pkg/state/appender.go [hunk 2]
```diff
@@ -641,16 +670,16 @@ func (a *txAppender) appendBlock(params *appendBlockParams) error {
...
 	}
-	// a temporary dummy for linters
-	if len(initialSnapshot) > 100 {
-		zap.S().Debug(initialSnapshot)
-	}
-	// Save miner diff first.
-	if err := a.diffStor.saveTxDiff(minerAndRewardDiff); err != nil {
+
+	// TODO apply this snapshot when balances are refatored
+	// err = initialSnapshot.Apply(&snapshotApplier)
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/appender.go:679]: It is

File: pkg/state/transaction_differ_test.go [hunk 3]
```diff
@@ -700,7 +738,7 @@ func TestCreateDiffExchangeV3WithProofsWithMixedOrders(t *testing.T) {
 // and produces an incorrect or unexpected diff, should be fixes some how
 //
 //	func TestCreateDiffExchangeWithSignature(t *testing.T) {
-//		to, path := createDifferTestObjects(t)
+//		to, path := createDifferTestObjects(t, checkerInfo
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/transaction_differ_test.go:742]: This function is used for tests. It initializes transaction performer, and transaction performer needs checkerInfo now to initialize snapshot applier

File: pkg/state/snapshot_generator.go [hunk 4]
```diff
@@ -365,214 +402,384 @@ func (sg *snapshotGenerator) generateSnapshotForUpdateAssetInfoTx(assetID crypto
...
+	if senderDataEntries == nil {
+		return errors.New("senderDataEntries map is not initialized")
+	}
+	if _, ok := senderDataEntries[sender]; ok {
+		entries := senderDataEntries[sender]
+		entries = append(entries, action.Entry)
+		senderDataEntries[sender] = entries
+	} else {
+		senderDataEntries[sender] = proto.DataEntries{action.Entry}
+	}
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/snapshot_generator.go]: senderDataEntries is a map, not a slice

File: pkg/state/snapshot_generator.go [hunk 5]
```diff
@@ -365,214 +402,384 @@ func (sg *snapshotGenerator) generateSnapshotForUpdateAssetInfoTx(assetID crypto
...
+	return nil
+}
+
+func (sg *snapshotGenerator) updateBalanceDiffFromPaymentAction(
+	action proto.AttachedPaymentScriptAction,
+	wavesBalanceDiff addressWavesBalanceDiff,
+	assetBalanceDiff addressAssetBalanceDiff,
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/snapshot_generator.go]: Not relevant anymore

File: pkg/state/snapshot_generator.go [hunk 6]
```diff
@@ -365,214 +402,384 @@ func (sg *snapshotGenerator) generateSnapshotForUpdateAssetInfoTx(assetID crypto
...
+	issuer proto.WavesAddress) ([]proto.AtomicSnapshot, error) {
+	var atomicSnapshots []proto.AtomicSnapshot
+	var assetInf *assetInfo
+	assetInf, err := sg.stor.assets.newestAssetInfo(proto.AssetIDFromDigest(action.AssetID))
+	if err != nil {
+		return nil, err
+	}
+	quantityDiff := big.NewInt(action.Quantity)
+	resQuantity := assetInf.quantity.Sub(&assetInf.quantity, quantityDiff)
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/snapshot_generator.go]: assetInf.quanity has a big.Int type, i wouldn't risk trying to convert a big int number to an int64. I think big.Int is there for a reason

File: pkg/state/snapshot_generator.go [hunk 7]
```diff
@@ -365,214 +402,384 @@ func (sg *snapshotGenerator) generateSnapshotForUpdateAssetInfoTx(assetID crypto
...
-				if err != nil {
-					return nil, errors.Wrap(err, "failed to generate snapshots for a lease cancel action")
-				}
-				snapshot = append(snapshot, leaseStatusSnapshot, senderLeaseBalanceSnapshot, recipientLeaseBalanceSnapshot)
-			default:
-				return nil, errors.Errorf("unknown script action type %T", a)
-			}
+	var amount = -int64(leasingInfo.Amount)
+	leaseStatusSnapshot, senderLeaseBalanceSnapshot, recipientLeaseBalanceSnapshot, err :=
+		sg.generateLeaseAtomicSnapshots(action.LeaseID, *leasingInfo, txID, leasingInfo.Sender, leasingInfo.Recipient, amount)
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/snapshot_generator.go]: Not relevant anymore, but also, the validation s in transaction checker

File: pkg/state/appender.go [hunk 8]
```diff
@@ -337,8 +339,7 @@ func (a *txAppender) commitTxApplication(
 	// Update complexity.
 	a.sc.addRecentTxComplexity()
 	// Save balance diff.
-	// TODO get balances snapshots
-	if err := a.diffStor.saveTxDiff(applicationRes.changes.diff); err != nil {
+	if err = a.diffStor.saveTxDiff(applicationRes.changes.diff); err != nil {
```
[the remark this answers is not in the document]
Comment (esuwu) [reply pkg/state/appender.go:342]: it is not relevant to this PR

**Pull request merged**